<a href="https://colab.research.google.com/github/hananfarizta/Python_Tugas4_StartupCampus/blob/main/04_Cupcake_Muhammad_Hanan_Rafif_Farizta_(HW01).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Preparation

Preparation biasa dilakukan untuk mempersiapkan data sebelum masuk dalam tahap pemodelan. <br>
Berikut adalah tahapan yang akan dilalui pada data `SC_HW1_bank_data.csv` sebelum tahap pemodelan :
1. Import Library
2. Input Dataset
3. Preprocessing
4. Train-Test Split

## Import Library

In [28]:
import pandas as pd
import numpy as np

## Input Dataset

In [29]:
#Membaca data dan memasukkannya ke dalam bentuk Pandas Dataframe
df = pd.read_csv('https://raw.githubusercontent.com/Rietaros/kampus_merdeka/main/SC_HW1_bank_data.csv')

In [30]:
#Jalankan code untuk mengecek nama kolom yang tersedia
df.columns

Index(['RowNumber', 'CustomerId', 'Geography', 'Gender', 'Age', 'Tenure',
       'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember',
       'EstimatedSalary', 'Exited'],
      dtype='object')

In [31]:
df.head()

RowNumber  CustomerId Geography  Gender  Age  Tenure    Balance  \
0          1    15634602    France  Female   42       2       0.00   
1          2    15647311     Spain  Female   41       1   83807.86   
2          3    15619304    France  Female   42       8  159660.80   
3          4    15701354    France  Female   39       1       0.00   
4          5    15737888     Spain  Female   43       2  125510.82   

   NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0              1          1               1        101348.88       1  
1              1          0               1        112542.58       0  
2              3          1               0        113931.57       1  
3              2          0               0         93826.63       0  
4              1          1               1         79084.10       0

In [32]:
#Hilangkan kolom yang dirasa tidak relevan dengan model (contoh: ID). None dapat diisi dengan nama-nama kolom yang akan digunakan.
#Contoh df = df[['X1','X2', 'Y']].copy()

#START CODE
df = df[['Geography', 'Gender', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember', 'EstimatedSalary', 'Exited']].copy()

In [33]:
df.head()

Geography  Gender  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0    France  Female   42       2       0.00              1          1   
1     Spain  Female   41       1   83807.86              1          0   
2    France  Female   42       8  159660.80              3          1   
3    France  Female   39       1       0.00              2          0   
4     Spain  Female   43       2  125510.82              1          1   

   IsActiveMember  EstimatedSalary  Exited  
0               1        101348.88       1  
1               1        112542.58       0  
2               0        113931.57       1  
3               0         93826.63       0  
4               1         79084.10       0

## Preprocessing

In [34]:
#Lakukan One-Hot Encoder untuk data categorical, dengan fungsi pandas get_dummies

#START CODE
df_one_hot = pd.get_dummies(df, prefix=['Geography', 'Gender'])

In [35]:
df_one_hot

Age  Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0      42       2       0.00              1          1               1   
1      41       1   83807.86              1          0               1   
2      42       8  159660.80              3          1               0   
3      39       1       0.00              2          0               0   
4      43       2  125510.82              1          1               1   
...   ...     ...        ...            ...        ...             ...   
9995   39       5       0.00              2          1               0   
9996   35      10   57369.61              1          1               1   
9997   36       7       0.00              1          0               1   
9998   42       3   75075.31              2          1               0   
9999   28       4  130142.79              1          1               0   

      EstimatedSalary  Exited  Geography_France  Geography_Germany  \
0           101348.88       1                 1                  0   
1           112542.58       0                 0                  0   
2           113931.57       1                 1                  0   
3            93826.63       0                 1                  0   
4            79084.10       0                 0                  0   
...               ...     ...               ...                ...   
9995         96270.64       0                 1                  0   
9996        101699.77       0                 1                  0   
9997         42085.58       1                 1                  0   
9998         92888.52       1                 0                  1   
9999         38190.78       0                 1                  0   

      Geography_Spain  Gender_Female  Gender_Male  
0                   0              1            0  
1                   1              1            0  
2                   0              1            0  
3                   0              1            0  
4                   1              1            0  
...               ...            ...          ...  
9995                0              0            1  
9996                0              0            1  
9997                0              1            0  
9998                0              0            1  
9999                0              1            0  

[10000 rows x 13 columns]

In [36]:
#Pisahkan mana X (feature) dengan Y,
#Y adalah kolom "Exited"

#START CODE
X = df_one_hot.drop("Exited", axis=1)
y = df_one_hot['Exited']

In [37]:
#Lakukan Scaler dan/atau Noermalisasi jika diperlukan
from sklearn.preprocessing import MinMaxScaler

#START CODE
scaler = MinMaxScaler()
X_transform = scaler.fit_transform(X)

In [38]:
#Ini digunakan jika dilakukan scaler/Normalisasi. Jika tidak, code ini bisa dilewat dan diganti dengan code yang ada di dalam komen
X_transform = pd.DataFrame(X_transform, columns = X.columns)
#X_transform = X.copy()

In [39]:
X_transform

Age  Tenure   Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     0.324324     0.2  0.000000       0.000000        1.0             1.0   
1     0.310811     0.1  0.334031       0.000000        0.0             1.0   
2     0.324324     0.8  0.636357       0.666667        1.0             0.0   
3     0.283784     0.1  0.000000       0.333333        0.0             0.0   
4     0.337838     0.2  0.500246       0.000000        1.0             1.0   
...        ...     ...       ...            ...        ...             ...   
9995  0.283784     0.5  0.000000       0.333333        1.0             0.0   
9996  0.229730     1.0  0.228657       0.000000        1.0             1.0   
9997  0.243243     0.7  0.000000       0.000000        0.0             1.0   
9998  0.324324     0.3  0.299226       0.333333        1.0             0.0   
9999  0.135135     0.4  0.518708       0.000000        1.0             0.0   

      EstimatedSalary  Geography_France  Geography_Germany  Geography_Spain  \
0            0.506735               1.0                0.0              0.0   
1            0.562709               0.0                0.0              1.0   
2            0.569654               1.0                0.0              0.0   
3            0.469120               1.0                0.0              0.0   
4            0.395400               0.0                0.0              1.0   
...               ...               ...                ...              ...   
9995         0.481341               1.0                0.0              0.0   
9996         0.508490               1.0                0.0              0.0   
9997         0.210390               1.0                0.0              0.0   
9998         0.464429               0.0                1.0              0.0   
9999         0.190914               1.0                0.0              0.0   

      Gender_Female  Gender_Male  
0               1.0          0.0  
1               1.0          0.0  
2               1.0          0.0  
3               1.0          0.0  
4               1.0          0.0  
...             ...          ...  
9995            0.0          1.0  
9996            0.0          1.0  
9997            1.0          0.0  
9998            0.0          1.0  
9999            1.0          0.0  

[10000 rows x 12 columns]

## Train-Test Split

In [40]:
#Split menjadi train dan test dengan test_size 25%
#Tidak perlu mengubah code ini

from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X_transform,y,test_size = 0.25,random_state = 123)

# Modeling

## Model1

### Soal :
Jelaskan secara Singkat Model pertama yang digunakan!

In [41]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

from sklearn.linear_model import LogisticRegression
model1 = LogisticRegression()
params = { 
    "tol": [0.1,0.01,0.001],
    'C':[0.5,1.0,1.5,2.0]
}


#START CODE
# model1 = None
# params = None
#END CODE

#Lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
from sklearn.model_selection import GridSearchCV
grid = GridSearchCV(
             estimator= model1,
             param_grid= params,
             scoring = 'accuracy',
             n_jobs = 10, # core cpu yang digunakan
             cv = 10 # 3-fold cross validation (artinya kita melakukan iterasi model sebanyak 3 kali)
            )

grid.fit(X_train,y_train)
grid.best_params_

{'C': 0.5, 'tol': 0.1}

In [42]:
grid.best_score_

0.8103999999999999

In [43]:
#lakukan evaluasi dengan beberapa maetric di bawah ini
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.82      0.97      0.89      1983
           1       0.66      0.20      0.31       517

    accuracy                           0.81      2500
   macro avg       0.74      0.59      0.60      2500
weighted avg       0.79      0.81      0.77      2500


[[1929   54]
 [ 412  105]]

0.8136


## Model2
### Soal :
Jelaskan secara Singkat Model ke-2 yang digunakan!

In [44]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

from sklearn.tree import DecisionTreeClassifier
model2 = DecisionTreeClassifier()
params2 = { 
    'criterion':['gini', 'entropy'],
    'max_depth':[2,4,6,8,10,12]
}


#START CODE
# model2 = None
# params = None
#END CODE

#Lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
from sklearn.model_selection import GridSearchCV
grid = GridSearchCV(
             estimator= model2,
             param_grid= params2,
             scoring = 'accuracy',
             n_jobs = 10, # core cpu yang digunakan
             cv = 10 # 3-fold cross validation (artinya kita melakukan iterasi model sebanyak 3 kali)
            )

grid.fit(X_train,y_train)
grid.best_params_

{'criterion': 'gini', 'max_depth': 6}

In [45]:
grid.best_score_

0.8578666666666667

In [46]:
#lakukan evaluasi
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.87      0.96      0.92      1983
           1       0.77      0.46      0.57       517

    accuracy                           0.86      2500
   macro avg       0.82      0.71      0.74      2500
weighted avg       0.85      0.86      0.84      2500


[[1911   72]
 [ 281  236]]

0.8588


## Model3
### Soal :
Jelaskan secara Singkat Model ke-3 yang digunakan!

In [47]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

from sklearn.ensemble import RandomForestClassifier
model3 = RandomForestClassifier()
params3 = { 
    'criterion':['gini', 'entropy'],
    'max_depth':[2,4,6,8,10,12]
}


#START CODE
# model3 = None
# params = None
#END CODE

#Lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
from sklearn.model_selection import GridSearchCV
grid = GridSearchCV(
             estimator= model3,
             param_grid= params3,
             scoring = 'accuracy',
             n_jobs = 10, # core cpu yang digunakan
             cv = 10 # 3-fold cross validation (artinya kita melakukan iterasi model sebanyak 3 kali)
            )

grid.fit(X_train,y_train)
grid.best_params_

{'criterion': 'entropy', 'max_depth': 12}

In [48]:
grid.best_score_

0.8633333333333333

In [49]:
#lakukan evaluasi
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.87      0.97      0.92      1983
           1       0.79      0.45      0.58       517

    accuracy                           0.86      2500
   macro avg       0.83      0.71      0.75      2500
weighted avg       0.86      0.86      0.85      2500


[[1921   62]
 [ 282  235]]

0.8624


## Tarik Kesimpulan Model Mana yang terbaik beserta alasannya

Menurut saya dari 3 model yang saya buat yang merupakan model terbaik yaitu menggunakan metode RandomForestClassifier karena menghasilkan tingkat keakuratan sebesar 0.86 atau 86% yang dimana lebih tinggi dibanding 2 metode lainnya yaitu DecisionTreeClassifier dan LogisticRegression